src/Extensions/Compatability/Validation/Code.cs

```diff
@@ -0,0 +1,515 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Omex.Extensions.Compatability.Validation
+{
+	/// <summary>
+	/// Code validation
+	/// </summary>
+	public static class Code
+	{
+		/// <summary>
+		/// Check object's state, throws specified Exception type if state evaluates to false.
+		/// </summary>
+		/// <typeparam name="TException">Exception to throw</typeparam>
+		/// <param name="state">Object's state</param>
+		/// <param name="errorMessage">Error message</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static void Expects<TException>(bool state, string errorMessage, EventId? tagId)
+			where TException : Exception
+		{
+			if (!Validate(state, errorMessage, tagId))
+			{
+				ReportError<TException>(errorMessage);
+			}
+		}
+
+
+		/// <summary>
+		/// Checks that the enumerable argument is not null and doesn't contain any nulls
+		/// </summary>
+		/// <remarks>Be careful to not pass enumerables that can be enumerated only once</remarks>
+		/// <typeparam name="TEnumerable">Type of the enumerable</typeparam>
+		/// <param name="argumentValue">The argument value.</param>
+		/// <param name="argumentName">Name of the argument.</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <exception cref="ArgumentException">Thrown if any argument  <paramref name="argumentValue"/> element is null.</exception>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static IEnumerable<TEnumerable> ExpectsAllNotNull<TEnumerable>([ValidatedNotNull] IEnumerable<TEnumerable>? argumentValue, string argumentName, EventId? tagId)
```

Comment:
Missing return type documentation throughout the changelist

Reply:
explained here https://github.com/microsoft/Omex/pull/153#discussion_r390509059